multicluster/controllers/multicluster/leader_clusterset_controller.go

```diff
@@ -0,0 +1,174 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package multicluster
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+
+	"go.uber.org/multierr"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	multiclusterv1alpha1 "antrea.io/antrea/multicluster/apis/multicluster/v1alpha1"
+	"antrea.io/antrea/multicluster/controllers/multicluster/common"
+	"antrea.io/antrea/multicluster/controllers/multicluster/core"
+)
+
+// LeaderClusterSetReconciler reconciles a ClusterSet object
+// There will be one controller running in each namespace of the leader for multiple cluster-set support
+// so each controller will only be handling a single cluster set in the given namespace.
+// TODO: Split the reconciler for member and leader to avoid if checks. In case a cluster is both a
+//       member and a leader, 2 instances of the controller will need to be run, one as a leader
+//		 the second as a member.
+type LeaderClusterSetReconciler struct {
+	client.Client
+	mutex  sync.Mutex
+	Scheme *runtime.Scheme
+
+	clusterSetConfig *multiclusterv1alpha1.ClusterSet
+	clusterSetID     common.ClusterSetID
+	clusterID        common.ClusterID
+
+	LocalClusterManager core.LocalClusterManager
+}
+
+//+kubebuilder:rbac:groups=multicluster.crd.antrea.io,resources=clustersets,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=multicluster.crd.antrea.io,resources=clustersets/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=multicluster.crd.antrea.io,resources=clustersets/finalizers,verbs=update
+
+// Reconcile ClusterSet CRD changes
+func (r *LeaderClusterSetReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	clusterSet := &multiclusterv1alpha1.ClusterSet{}
+	err := r.Get(ctx, req.NamespacedName, clusterSet)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			return ctrl.Result{}, err
+		}
+		// if errors.IsNotFound(err)
+		klog.InfoS("Received ClusterSet delete", "config", klog.KObj(clusterSet))
+		var stopErr, retErr error = nil, nil
+		if stopErr = r.LocalClusterManager.Stop(); stopErr != nil {
+			multierr.Append(retErr, stopErr)
+		}
+		r.LocalClusterManager = nil
+		r.clusterSetConfig = nil
+		r.clusterID = common.INVALID_CLUSTER_ID
+		r.clusterSetID = common.INVALID_CLUSTER_SET_ID
+
+		return ctrl.Result{}, stopErr
+	}
+
+	klog.InfoS("Received ClusterSet add/update", "config", klog.KObj(clusterSet))
+
+	// Handle create or update
+	// If create on leader, make sure the local cluster claim is part of the leader config
+	if err = r.validateLocalClusterClaim(clusterSet); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// if update, leader has nothing to do in inbound mode
```

Comment:
remove `in inbound mode` since there is no context?

Reply:
Yes I removed it in the latest patch.